### PR TITLE
BUG: Allow x to be just increasing instead of strictly increasing

### PR DESCRIPTION
**Pull request by TimothyWillard**
Allow x to be just increasing instead of strictly increasing for scipy.interpolate's UnivariateSpline, InterpolatedUnivariateSpline, LSQUnivariateSpline. See ticket #8535.

### CODE REVIEW

File: scipy/interpolate/fitpack2.py [hunk 1]
```diff
@@ -543,7 +543,8 @@ class InterpolatedUnivariateSpline(UnivariateSpline):
         not in the interval defined by the knot sequence.
 
         * if ext=0 or 'extrapolate', return the extrapolated value.
-        * if ext=1 or 'zeros', return 0
+        * if ext=1 all(diff(x) > 0.0):
++            raise ValueEror 'zeros', return 0
```
Comment (pv) [inline scipy/interpolate/fitpack2.py]: This is not valid [restructuredtext](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) markup. Also not the strange `+` at the start of the line.

Comment (TimothyWillard) [reply scipy/interpolate/fitpack2.py]: Thanks so much! I didn't even notice that.

File: scipy/interpolate/fitpack2.py [hunk 2]
```diff
@@ -172,8 +172,8 @@ def __init__(self, x, y, w=None, bbox=[None]*2, k=3, s=None,
             if (not np.isfinite(x).all() or not np.isfinite(y).all() or
                     not w_finite):
                 raise ValueError("x and y array must not contain NaNs or infs.")
-        if not all(diff(x) > 0.0):
-            raise ValueError('x must be strictly increasing')
+        if not all(diff(x) >= 0.0):
+            raise ValueError("x must be increasing")
```
Comment (ilayn) [inline scipy/interpolate/fitpack2.py]: maybe `nondecreasing` is a better word here?

File: scipy/interpolate/tests/test_fitpack2.py [hunk 3]
```diff
@@ -177,6 +178,21 @@ def test_increasing_x(self):
...
+    def test_increasing_x(self):
+        # Test that x is required to be increasing, see ticket #8535
+        xx = np.arange(10, dtype=float)
+        yy = xx**3
+        x = np.arange(10, dtype=float)
+        x[1] = x[0] - 1.0
+        y = x**3
+        w = np.ones_like(x)
+        # also test LSQUnivariateSpline [which needs explicit knots]
+        spl = UnivariateSpline(xx, yy, check_finite=True)
```
Comment (ilayn) [inline scipy/interpolate/tests/test_fitpack2.py]: You don't seem to be using x, y, w for anything special. Maybe you can reuse `xx`, `yy`, `w` here ?

File: scipy/interpolate/fitpack2.py (hunk 2 above)
Comment (jeffyancey) [reply scipy/interpolate/fitpack2.py]: I like this better as well.

**Comment by andyfaff** [comment]
Sorry to be late to the party here, but I'd like to know why it's not a requirement of `UnivariateSpline` for `x` to be strictly increasing? What does `UnivariateSpline` do, or what should it be expected to do, for multivalued functions, such as Case 1 below? For example:

```
# Case 1: perhaps a problem, what does `UnivariateSpline` do?
x = np.array([0., 1, 2, 2, 3])
y = np.array([0., 1, 2, 3, 3])

# Case 2: shouldn't create a problem
x = np.array([0., 1, 2, 2, 3])
y = np.array([0., 1, 2, 2, 3])
```

I went back to #1753, but that didn't clear anything up for me. Case 2 is the kind of thing that should be legal usage. Case 1 would exist in some problems. @charris, could you illuminate what should happen? Would the code just do some kind of average?
Could we discuss this before the PR is merged? Either way I think the documentation can be improved here. Either saying why strictly increasing `x` is required, or what the interpolation does in cases of multivalued functions. Perhaps this discussion is relevant to all interpolation functions?

**Comment by ev-br** [comment]
Also the behavior should be identical for UnivariateSpline and splrep/splev, best add a corresponding test.

**Comment by pv** [comment]
@andyfaff: for general `s`, FITPACK returns least-squares smoothing spline so in principle it is not problem to have multiple values. However, in the interpolation limit `s->0` the fitting problem seems to become ill-conditioned for Case 2, and nonsense is returned.

**Comment by ggrrll** [comment]
@pv  @andyfaff indeed, when I use
```
if np.any(dx < 0):
            raise ValueError("`x` must be non decreasing sequence.")
```
I get this error message , so... I don't think the solution is so easy ...
screen shot 2018-10-12 at 15 41 16

would be maybe useful to see how this is implemented in R
[smooth.spline](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/smooth.spline.html) with 
[gam method](https://stat.ethz.ch/R-manual/R-devel/library/mgcv/html/gam.html)

**Comment by jxrossel** [comment]
Could we imagine to require that x is strictly increasing for interpolation cases (s=0) and non-strictly increasing for smoothing cases?

**Comment by ev-br** [comment]
Superseded by gh-10608. Thanks all

**Pull request closed without merging**